src/lib/storage/dictionary_segment.cpp

```diff
@@ -0,0 +1,133 @@
+#include <unordered_set>
+
+#include "dictionary_segment.hpp"
+#include "type_cast.hpp"
+
+#include "utils/assert.hpp"
+
+namespace opossum {
+
+template <typename T>
+DictionarySegment<T>::DictionarySegment(const std::shared_ptr<AbstractSegment>& abstract_segment) {
+  for (auto index = ChunkOffset{0}; index < abstract_segment->size(); index++) {
+    const auto& value = type_cast<T>(abstract_segment->operator[](index));
+    _dictionary.push_back(value);
+  }
+
+  // use standard library functions to remove duplicate values from dictionary
+  // sorted dictionary also allows us to be more efficient in element lookup
+  std::sort(begin(_dictionary), end(_dictionary));
+  const auto& end_it = std::unique(begin(_dictionary), end(_dictionary));
+  _dictionary.erase(end_it, _dictionary.end());
+  _dictionary.shrink_to_fit();
+
+  // determine how many bits are needed to encode based on dictionary size
+  const auto num_bits = std::ceil(std::log2(_dictionary.size()));
+  Assert(num_bits <= 32, "The dictionary is too large for this compression algorithm!");
+
+  // initialize attribute vector with smallest applicable integer type
+  if (num_bits <= 8) {
+    _attribute_vector = std::make_shared<FixedWidthIntegerVector<uint8_t>>();
+  } else if (num_bits <= 16) {
+    _attribute_vector = std::make_shared<FixedWidthIntegerVector<uint16_t>>();
+  } else {
+    _attribute_vector = std::make_shared<FixedWidthIntegerVector<uint32_t>>();
+  }
```

Comment:
I would add an assertion for the else case. This could help to migrate to uint64_t in the future